hivemq-edge/src/frontend/src/modules/Workspace/utils/adapter.utils.ts

```diff
@@ -0,0 +1,43 @@
+import { ProtocolAdapter } from '@/api/__generated__'
+import { IconType } from 'react-icons'
+import { TbSettingsAutomation } from 'react-icons/tb'
+import { FaIndustry } from 'react-icons/fa6'
+import { GrConnectivity } from 'react-icons/gr'
+import { AiFillExperiment } from 'react-icons/ai'
+import { MdInput, MdOutlineFindInPage, MdOutput } from 'react-icons/md'
+
+/**
+ * @deprecafted This is a mock, replacing the missing WRITE capability from the adapters
+ * @param adapter Adapter | undefined
+ */
+export const isBidirectional = (adapter: ProtocolAdapter | undefined) => {
+  return Boolean(adapter?.id?.includes('opc-ua-client'))
+}
+
+/**
+ * @deprecated This is a mock, mapping should be based on ProtocolAdapterCategory and image property
+ * @see ProtocolAdapterCategory
+ */
+export const deviceCategoryIcon: Record<string, IconType> = {
+  BUILDING_AUTOMATION: TbSettingsAutomation,
+  INDUSTRIAL: FaIndustry,
+  CONNECTIVITY: GrConnectivity,
+  SIMULATION: AiFillExperiment,
+}
+
+// https://stackoverflow.com/questions/41253310/typescript-retrieve-element-type-information-from-array-type
+type ArrayElement<ArrayType extends readonly unknown[]> = ArrayType extends readonly (infer ElementType)[]
+  ? ElementType
+  : never
```

Comment:
suggestion(non-blocking): I think this is an old solution which can be done "easier" in the latest TypeScript versions:

```ts
type ArrayElement<ArrayType extends Array> = ArrayType[number]
//                                                    ^ this is the important element, as it's not an accessor but the definition that you want an inner element
```

Alternately you can also do the apply the same principles to direct types: E.g.: on `CapabilitiesArray`

```ts
type CapabilityType = CapabilitiesArray[number] | 'WRITE'
```